### ISSUE TITLE
Installation Difficulties 

### ISSUE THREAD

**Issue by MozhganPourKeshavarz** [thm-projects/arsnova-backend#68]
Hello there,

I have tried to install ARSnova on an Ubuntu 18 system with no success so far. I have installed arsnova-backend and I can verify the installation using the following command:

```
$ curl -H "Accept: application/json" localhost:8080/api/
{"version":{"string":"2.7.0","buildTime":"2019-01-31T18:59:42Z","gitDirty":false,"gitCommitId":"bcb378a59342ff76693689cd13e924a7f68e5135"},"productName
```

As for the frontend, I have installed arsnova-mobile which is verified below:

```
$ curl localhost:8080/mobile/resources/version.json
{"productName":"arsnova-mobile","version":{"string":"2.7.0","gitCommitId":"ebd75e8ca7342c4053cad9064aaee01548608246","gitDirty":false,"buildTime":"2019-02-01T15:48:47.392Z"}}
```

However, in the Tomcat Web Application Manager if I go to either of `/api` or `/mobile` the system gives the same Notice that `The connection to the server has failed!`

When looking at the console for the HTML code generated, it gives the following error:

```
Could not load configuration
(anonymous) @ VM28:1
```

I appreciate your help with this!
Mozhgan

**Comment by commana** [comment]
Hi @MozhganPourKeshavarz,

Do you see any suspicious log output from arsnova-backend?

Not sure if this is related to your problem, but: the documentation is outdated at the moment concerning the `arsnova.properties` file. This is now a YAML file named `defaults.yml` in the resource folder that needs to be copied to /etc/arsnova/application.yml in order to make changes.

Hope this helps!

**Comment by MozhganPourKeshavarz** [comment]
@commana  Thanks for your fast reply,
we are using ARSnova version 2.7(as you mention switch to branch 2.x for stable version), So the file name "arsnova.properties" is correct,
we have installed the following packages on Ubunto 18:
Erlang R16B03(build from src)
Apache CouchDB 1.5(build from src) (it works fine independently)
Apache Tomcat 8
Apache HTTP Server 

as I said,when I go to either of /api or /mobile in the Tomcat Web Application Manager the system gives the same Notice that The connection to the server has failed! and some errors in browser inspect element as bellow:

![Screenshot from 2019-09-02 19-44-47](https://user-images.githubusercontent.com/36272225/64123811-9477c880-cdba-11e9-8f2c-ad34f8fc4102.png)

we also checked tomcat8 logfile and it seems that there is a problem in arsnova-config path as you can see:
```
...
127.0.0.1 - - [02/Sep/2019:19:30:20 +0430] "GET /mobile/resources/css/index.css HTTP/1.1" 304 -
127.0.0.1 - - [02/Sep/2019:19:30:20 +0430] "GET /arsnova-config HTTP/1.1" 404 1097
127.0.0.1 - - [02/Sep/2019:19:30:21 +0430] "GET /arsnova-config?_dc=1567436421250 HTTP/1.1" 404 1097
127.0.0.1 - - [02/Sep/2019:19:30:21 +0430] "GET /mobile/app.json?1567436421269 HTTP/1.1" 200 1275
127.0.0.1 - - [02/Sep/2019:19:30:21 +0430] "GET /mobile/app.js?1567436421316 HTTP/1.1" 200 1507283
```
my Virtualhost section in `/etc/apache2/sites-available/arsnova.conf` looks now like this:
```
		RewriteEngine On
		RewriteRule ^/$  http://127.0.0.1:8080/mobile/ [R=301,L]
		RewriteRule ^/arsnova-config$ /api/configuration/ [PT,L]
		<Location /api/>
			ProxyPass http://127.0.0.1:8080/api/
			ProxyPassReverse http://127.0.0.1:8080/api/
			ProxyPreserveHost On
			RequestHeader set X-Forwarded-Proto %{REQUEST_SCHEME}
		</Location>
		<Location /socket.io/>
			ProxyPass ws://127.0.0.1:8090/socket.io/
			ProxyPassReverse ws://127.0.0.1:8090/socket.io/
		</Location>

		<Location /mobile/>
			ProxyPass http://127.0.0.1:8080/mobile/
			ProxyPassReverse http://127.0.0.1:8080/mobile/
		</Location>

		<Location /arsnova-config>
		ProxyPass http://127.0.0.1:8080/api/arsnova-config
		ProxyPassReverse http://127.0.0.1:8080/api/arsnova-config
	</Location>

		Redirect / http://arsnova.rubel.rub.de/mobile/
```

So have you any idea?
Thanks.

**Comment by dgerhardt** [comment]
Hi @MozhganPourKeshavarz,

your browser screenshots shows accessing ARSnova directly over port 8080 so the proxy is not used. This doesn't work because paths to access the API differ in this case. You have to access ARSnova through the proxy port, i.e. port 80 or 443 with HTTPS.

Another issue might be caused by the `<Location /arsnova-config>` block. This block should not be necessary.

**Comment by MozhganPourKeshavarz** [comment]
Hi @dgerhardt,

Thanks for your response. We tried to look into Apache setup to make sure proxy is set up properly with no success so far.

In order to be able to debug the system, we would like to test Tomcat8 independently without using any proxy. Would you be able to help us with that? That is, how do we directly test Tomcat8? Right now, for example, if we use http://127.0.0.1:8080/mobile/# it gives an error: "The connection to the server has failed" (shown in the screenshot in my previous message).

In short, what is an appropriate way of testing the Tomcat8 part of the system? 

Thanks for your help again.

**Comment by tekay** [comment]
HI @MozhganPourKeshavarz,
also joining in on this :).

Can you have a look at the `/var/log/tomcat8/catalina.out` ?

**Comment by MozhganPourKeshavarz** [comment]
@tekay  Hi there,
this is all you wanted : 
[catalina.out.txt](https://github.com/thm-projects/arsnova-backend/files/3686920/catalina.out.txt)
could you let me some hints?

**Comment by dgerhardt** [comment]
It looks like the database isn't running:
```
18:40:44,325  INFO pool-2-thread-1 dao.CouchDBDao:544 - Initializing database connection.
18:40:44,328 ERROR pool-2-thread-1 couchdb.Session:475 - org.apache.http.conn.HttpHostConnectException: Connect to localhost:5984 [localhost/127.0.0.1] failed: Connection refused (Connection refused)
```

Please make sure that all dependencies described in https://github.com/thm-projects/arsnova-backend/blob/master/src/site/markdown/installation.md#server-software are installed.

Regarding running ARSnova without reverse proxy:
It's possible but I don't advice it. You need different ARSnova settings w/ and w/o proxy and there isn't much that can go wrong with the proxy setup. But if you still want to try it:
* Rename api.war in Tomcat's webapp folder to ROOT.war.
* Make sure that `socketio.proxy-path=/socket.io` is commented out.
* Once you use a proxy, `socketio.proxy-path` has to be set and the .war file need to be renamed back to api.war.

**Comment by MozhganPourKeshavarz** [comment]
@dgerhardt 
Thanks for the great hints. Now it is almost working. 
It now works from the `localhost:8080`
However it does not work from the `serverip:8080` which means the system is not accessible from other machines on the network.

After a fresh install, we make ONLY the following changes:
- Renamed `api.war` in Tomcat's webapp folder to `ROOT.war`
- In the file `/etc/appache2/sites-available/arsnova.conf` we commented out the following line:
` RequestHeader set X-Forwarded-Proto %{REQUEST_SCHEME}`

With these changes the system works on the localhost.

It seems like we can keep the first change, renaming api to ROOT, because we are not planning to use any other tomcat webapp on the same server.

However, the Appache line above seems to be important for production machine.

The problem is that if we uncomment the `RequestHeader` line, then Appache gives the following error when restarting:

```
hmaleki@E151:~$ sudo systemctl restart apache2
Job for apache2.service failed because the control process exited with error code.
See "systemctl status apache2.service" and "journalctl -xe" for details.
hmaleki@E151:~$ sudo systemctl status apache2
● apache2.service - The Apache HTTP Server
   Loaded: loaded (/lib/systemd/system/apache2.service; enabled; vendor preset: enabled)
  Drop-In: /lib/systemd/system/apache2.service.d
           └─apache2-systemd.conf
   Active: failed (Result: exit-code) since Sat 2019-10-05 12:36:07 +0330; 3s ago
  Process: 21734 ExecStop=/usr/sbin/apachectl stop (code=exited, status=1/FAILURE)
  Process: 25965 ExecReload=/usr/sbin/apachectl graceful (code=exited, status=0/SUCCESS)
  Process: 21740 ExecStart=/usr/sbin/apachectl start (code=exited, status=1/FAILURE)
 Main PID: 18977 (code=exited, status=0/SUCCESS)

اوْکتوْبر 05 12:36:07 E151 systemd[1]: Starting The Apache HTTP Server...
اوْکتوْبر 05 12:36:07 E151 apachectl[21740]: AH00526: Syntax error on line 44 of /etc/apache2/sites-enabled/arsnova.conf:
اوْکتوْبر 05 12:36:07 E151 apachectl[21740]: Invalid command 'RequestHeader', perhaps misspelled or defined by a module not included in the server config
اوْکتوْبر 05 12:36:07 E151 apachectl[21740]: Action 'start' failed.
اوْکتوْبر 05 12:36:07 E151 apachectl[21740]: The Apache error log may have more information.
اوْکتوْبر 05 12:36:07 E151 systemd[1]: apache2.service: Control process exited, code=exited status=1
اوْکتوْبر 05 12:36:07 E151 systemd[1]: apache2.service: Failed with result 'exit-code'.
اوْکتوْبر 05 12:36:07 E151 systemd[1]: Failed to start The Apache HTTP Server.

```

**Comment by dgerhardt** [comment]
Regarding `RequestHeader`:
`a2enmod headers` should help. It looks like Ubuntu 18.04 doesn't enable this module by default. I'll add it to the install guide.

> It now works from the localhost:8080
> However it does not work from the serverip:8080

You probably followed this section and have to temporarily remove the `address` attribute:

> By default, Tomcat listens on a public interface. If you follow our recommendation to use a reverse proxy, you should change the configuration so Tomcat only accepts requests from localhost. Open /etc/tomcat9/server.xml, look for <Connector port="8080" ...>, and change it to:
>
>     <Connector port="8080" address="127.0.0.1" ...>

**Comment by MozhganPourKeshavarz** [comment]
@dgerhardt  It works fine now and serverip:8080 is reachable from other machine,
but we still have a problem in reverse proxy, 
now we set `RequestHeader` in the file  `/etc/appache2/sites-available/arsnova.conf` with a small change and there is no error.
changed version : `RequestHeader set "X-Forwarded-Proto"  expr=%{REQUEST_SCHEME} `
have you any idea to solve that?
thanks.
as you mentioned we tried to enable a2enmod headers with these following commands:
```
sudo a2enmod headers
sudo service apache2 restart
```
with these configurations it works great, if we rename `api.war` file to `ROOT.war`, 
but when we rename back it to `api.war` it doesn't work.
there is no difference either activate or deactivate `socketio.proxy-path=/socket.io` in our case.

**Comment by dgerhardt** [comment]
> changed version : RequestHeader set "X-Forwarded-Proto" expr=%{REQUEST_SCHEME}

This should be fine.

> but when we rename back it to api.war it doesn't work.

That's not much information to work with. What doesn't work? The connection to Apache on port 80? Are there any errors in Apache's or Tomcat's logs? What are the HTTP responses on both ports?

    curl -v -H "Accept: application/json" localhost/api/
    curl -v -H "Accept: application/json" localhost:8080/api/

> there is no difference either activate or deactivate socketio.proxy-path=/socket.io in our case.

You'll notice that giving feedback doesn't work if this isn't set up correctly.

**Comment by MozhganPourKeshavarz** [comment]
@dgerhardt HTTP responses seem to be OK,
`curl -v -H "Accept: application/json" localhost/api/` :
```
*   Trying 127.0.0.1:80...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /api/ HTTP/1.1
> Host: localhost
> User-Agent: curl/7.65.2
> Accept: application/json
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 
< Date: Mon, 07 Oct 2019 14:45:23 GMT
< Server: Apache/2.4.29 (Ubuntu)
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
{"version":{"string":"2.7.1","buildTime":"2019-04-09T11:17:59Z","gitDirty":false,"gitCommitId":"a01af8c009fb500e3a9d7a9b8ea2b220b6c9f520"},"productName":"arsnova-backend"}
```
`
curl -v -H "Accept: application/json" localhost:8080/api/` :
```
*   Trying 127.0.0.1:8080...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /api/ HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.65.2
> Accept: application/json
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Mon, 07 Oct 2019 14:45:56 GMT
< 
* Connection #0 to host localhost left intact
{"version":{"string":"2.7.1","buildTime":"2019-04-09T11:17:59Z","gitDirty":false,"gitCommitId":"a01af8c009fb500e3a9d7a9b8ea2b220b6c9f520"},"productName":"arsnova-backend"}
```

when I set war file to api.war and go to either of /api or /mobile in the Tomcat Web Application Manager the system gives the same Notice that The connection to the server has failed! and some errors in browser inspect element as bellow : 
![Screenshot from 2019-09-02 19-44-47](https://user-images.githubusercontent.com/36272225/66322628-4a40b480-e92f-11e9-9a51-42d752a9f9f8.png)

but if I rename it back to ROOT.war and go to /mobile in the Tomcat Web Application Manager it works fine : 
![Screenshot from 2019-10-07 18-25-50](https://user-images.githubusercontent.com/36272225/66323336-832d5900-e930-11e9-9be8-a1f8ebad4fa2.png)
and in this case if I stop apache, it works fine too. so it means apache doesn't do anything!
and we are going to use proxypass, so what should we do?
thanks.

**Comment by MozhganPourKeshavarz** [comment]
One more question is how important the proxy is. In other word, what happens if the users are directly served by tomcat? For example, if there are about 80 people responding to a question, would tomcat be able to handle it, or having a proxy matters here? How about 500 people?

**Comment by dgerhardt** [comment]
> when I set war file to api.war and go to either of /api or /mobile in the Tomcat Web Application Manager the system gives the same Notice that The connection to the server has failed!

That's expected. If you don't use the proxy, the frontend cannot find the configuration below /api. With the proxy setup you have to access ARSnova through Apache's port which should be 80 by default w/o an SSL setup.

> One more question is how important the proxy is. In other word, what happens if the users are directly served by tomcat? For example, if there are about 80 people responding to a question, would tomcat be able to handle it, or having a proxy matters here? How about 500 people?

We haven't used Tomcat in production w/o a reverse proxy. But there are some known issues:
* Websockets use a different port which can cause problems with firewalls.
* Many things are harder to setup with Tomcat: e.g. SSL, HTTP/2 support, compression of static files.

**Comment by MozhganPourKeshavarz** [comment]
Thanks a lot @dgerhardt 
Given your response we want to go back to standard setup. 

1. For tomcat8, we use api.war, which is the default name. And we use the address `127.0.0.1:8080` for tomcat8
2. For Apache2, we use `VirtualHost 172.16.20.1:80`

However, we are having two problems with that setup. Specifically,
1. For tomcat8, again we get the old error message: "The connection to the server has failed!"
2. For Apache2, its really wired! When going to this address `172.16.20.1` it gets redirected to `127.0.0.1:8080`. What is even more amazing is that even if we shutdown Apache2 server, it still get forwarded (from `172.16.20.1` to `127.0.0.1:8080`)!!

**Comment by dgerhardt** [comment]
> When going to this address `172.16.20.1` it gets redirected to `127.0.0.1:8080`.

Check the `RewriteRule`s. You can use relative URIs here.

> if we shutdown Apache2 server, it still get forwarded (from `172.16.20.1` to `127.0.0.1:8080`)!!

That's probably the browser caching the redirection.
